Make hydra-explorer configurable

ffakenz opened the pull request:
<!-- Describe your change here -->

## Why
Soon we are going to host the hydra-explorer and want it to be listening on port 80.
Because in other environments port 80 might not be available, we need to make the explorer configurable.

## What
This PR aims to address the above.

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ffakenz committed:
Add options for hydra-explorer

ch1bo commented on hydra-explorer/src/Hydra/Explorer/Options.hs:
```diff
@@ -0,0 +1,41 @@
+optionsParser =
+  Options
+    <$> networkIdParser
+    <*> peerParser
```
Must: not use `--peer` to set the listen port of a web server.

The help text is irrelevant and this is confusing.

I think we should have a `--listen <host>:<port>` (or `--api-host` and `--api-port` like the hydra-node if you want to re-use things) or only a `--port <port>`.

Further, I think re-using the parsers (including help text and flags) from `hydra-node` is less clear than just duplicating the code to here.

ffakenz committed:
Refactor options to only take api port

ffakenz committed:
Fix the start from chain arg

This fixes the e2e explorer specs

v0d1ch commented on hydra-explorer/src/Hydra/Explorer.hs:
```diff
@@ -82,26 +84,33 @@ readModelGetHeadIds = readTVarIO
+          *> Warp.runSettings (settings tracer port) (httpApp tracer getHeads)
       )
       >>= \case
         Left{} -> error "Something went wrong"
```
Could we print the error here to get better insights into what happened? Perhaps also use dedicated exception type instead of `error`.

ffakenz replied:
the error here is of type Unit, so there is no information to be printed or type assertion to be made.

ch1bo replied:
Neither `Left` or `Right` is expected here. It would mean that either the chain observer stopped or the web server stopped.

Should: use `concurrently_` or to enforce both to run and rethrow exceptions if there are any.

Alternative would be to fail on both `Left` and `Right`, as neither is expected to return or you could also use `race_`.

ffakenz replied:
good catch!

going with `race_`

v0d1ch commented on hydra-explorer/src/Hydra/Explorer.hs:
```diff
@@ -82,26 +84,33 @@ readModelGetHeadIds = readTVarIO
 main = do
   withTracer (Verbose "hydra-explorer") $ \tracer -> do
+    opts <- execParser hydraExplorerOptions
+    let Options
```
These could be part of the `where` block as per our coding conventions :trollface:

ffakenz replied:
i dont see a way... opts is the result obtained from an action.

unless i am not understanding ur request.
can u plz elaborate more 🙏

v0d1ch replied:
Oh yeah, you are right

ffakenz committed:
Use race_ instead as there is no information to trace

ch1bo committed:
Build before running tests

This seems to result in smaller (build) closures.

As a next step we might not want to use dev shells for running these
tests.
